### PR TITLE
Implement functions to drop and interpolate over low-confidence datapoints

### LINKED ISSUE #97: Provide function for dropping values with low confidence and interpolating over them
## Background
Pose estimation tools usually provide confidence values for each predicted point.
For `DeepLabCut` and `LightningPose` this is named "likelihood", while for SLEAP this is known as (point-wise) confidence.
The exact metrics and their value ranges differ and do not translate well from one tool to another. Nevertheless, we store all of those in the `xarray.Datarray` named "confidence".

## What do we need
We now need to provide a way for users to:
1. Drop all points where the confidence is below a given threshold (i.e. set those points to NaN).
2. Interpolate over the resulting "gaps"

## Example implementation
A simple implementation of this workflow can be found in [this notebook](https://github.com/neuroinformatics-unit/course-behavioural-analysis/blob/main/notebooks/EPM_analysis_with_solutions.ipynb), and it relies on using two built-in `xarray` methods:

```python
confidence_threshold = 0.6
ds_thresholded = ds.where(ds.confidence > confidence_threshold)

# interpolate over missing values (NaNs) for gaps of up to max_gap seconds
ds_interp = ds_thresholded.interpolate_na(dim="time", method="cubic", max_gap=1)
```

The above (or some similar implementation) needs to be wrapped in a function that takes an `xarray.Dataset` as input and returns the processed `xarray.Dataset`. Ideally this would expose all "useful" parameters, such as `confidence_threshold`, `interolation_method`, and `max_gap`, and provide sensible default values for them. This may be tricky and will require some research, especially because the appropriate threshold value will depend on the `source_software` attribute.

## Some nice-to-haves
- provide an `inplace` parameter (similar to `pandas`), which modifies the dataset "in place" if `True`, but is `False` by default (i.e. the default behaviour is to create a new dataset, not to overwrite the old one).
- produce some diagnostic metrics that can be used for quality control (e.g. how many values were dropped, as an absolute number or as % of total values)
- Keep a record of this processing operation and its parameters somehow (e.g. store them in the resulting dataset's metadata?)

## Added context
It's okay to take a bit longer to get this one right, as it may serve as template for additional data cleaning / filtering methods.

The above is a rough sketch of how I envision this, no need to follow it closely. Feel free to discuss suggestions under this issue.

### PR DESCRIPTION
**Pull request by b-peri**
## Description

This PR introduces the `filtering` module, and adds two new major user-facing functions to `movement`. The first, `filter_by_confidence()`, drops pose track datapoints where the associated confidence value falls below a user-defined threshold, and replaces these with NaNs. The second, `interpolate_over_time()`, interpolates over such NaN values across the `time` axis using the xarray built-in `interpolate_na()`. 

_Edit 11/03/2024:_
This PR also introduces two new convenience functions for the `filtering` module:

1. The `log_to_attrs()` decorator adds a log of any operations performed (in the form of a dictionary) to the affected `xarray.Dataset`'s attributes, under `["log"]`.
2. The `filter_diagnostics()` function counts and logs the number of datapoints that have been filtered from every individual keypoint.

- [ ] Bug fix
- [X] Addition of a new feature
- [ ] Other

## References

Closes #97

## Checklist:

- [X] The code has been tested locally
- [X] Tests have been added to cover all new functionality
- [ ] The documentation has been updated to reflect any changes
- [X] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Comment by b-peri** [comment]
Some particular items I would like to receive some feedback on:
- Function names, particularly that of `interp_pose`. I think that a more generic function name (maybe even something like simply "interpolate") may be more informative, but given that the use-case is specifically for pose tracks in this case, I'm a bit unsure of how specific to go.
-  Both functions currently act both on `confidence` _and_ `pose_tracks` variables, but I'm unsure whether this makes sense (e.g. I don't think interpolating over confidence values is bound to be particularly informative). Shall I restrict `filter_confidence()` function to act only on the `pose_tracks` variable?
-  `filter_confidence()` currently reports some basic diagnostics, but I was unsure about what values to report exactly. I was also thinking we might like to allow users to optionally plot some confidence distributions + imposed thresholds. Any feedback on this would be highly appreciated!

File: movement/analysis/filtering.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+    # Diagnostics
+    print("\nDatapoints Filtered:\n")
+    for kp in ds.keypoints.values:
+        n_nans = np.count_nonzero(
+            np.isnan(ds_thresholded.confidence.sel(keypoints=f"{kp}").values)
+        )
+        n_points = ds.time.values.shape[0]
+        prop_nans = round((n_nans / n_points) * 100, 2)
+        print(f"{kp}: {n_nans}/{n_points} ({prop_nans}%)")
```
Comment (niksirbi) [inline movement/analysis/filtering.py:113]: This calculations don't work properly. For examply, on the "SLEAP_three-mice_Aeon" dataset I get:
```
Datapoints Filtered:

centroid: 1803/601 (300.0%)
```

There are several problems here:
- this dataset has NaN confidence values, in such cases we should not filter (so drop a value only if confidence is not None and is below the theshold)
- the 300% is obviously wrong, it counts all 3 centroids (from all 3 individuals). Perhaps it's better split the report by individual and by keypoint (instead of by keypoint only).

File: movement/analysis/filtering.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    # TODO: Is this enough diagnostics? Should I write logic to allow
+    #  users to optionally plot out confidence distributions + imposed
+    #  threshold?
```
Comment (niksirbi) [inline movement/analysis/filtering.py:117]: I would leave the plotting for another PR, the text report is good enough for now.

**Review by niksirbi — COMMENTED** [review]
Thanks a lot @b-peri! You did a great job, and opening this draft PR was very effective at laying out the problems and the solutions. Here are my high-level thoughts (details can be found in the individual comments):

- I like the logging, but I think we can generalise it to any filtering function via an appropriate decorator (see example decorator in the respective comment).
- Filtering should only operate on `pose_tracks`, and in the future perhaps on `velocity`, `acceleration` etc, definitely not on `confidence`
- Accordingly I propose renaming the functions as `interpolate_over_time()` and `filter_by_confidence()`.
- I think we should completely scrap the "inplace" idea (for reasons given in the comment). Make all functions return a new dataset.
- I think it's enough to have a text report of dropped values, plots can come later. But the current implementation of this feature is incorrect and should be fixed.
- For now, I would put the module in `movement/filtering.py`, not sure it counts as "analysis". When it grows, we can consider making it into a subfolder
- The interpolation makes use of the `bottleneck` library, which is not installed by default. The dependency needs to change to `xarray[accel]` instead of "vanilla" xarray (see [explanation](https://docs.xarray.dev/en/stable/getting-started-guide/installing.html#for-accelerating-xarray))

Let me know your thoughts on the above.

File: movement/analysis/filtering.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    if inplace:
+        ds["pose_tracks"] = ds_interpolated["pose_tracks"]
+        ds["confidence"] = ds_interpolated["confidence"]
+        ds.attrs["log"] = ds_interpolated.attrs["log"]
+        return None
+    else:
+        return ds_interpolated
+
+
+def filter_confidence(
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/analysis/filtering.py:71]: Also, currently you provide users with the option to run interpolation after the filtering, which is very convenient.
However, this can only be done with the default interpolation parameters, so I'm ambivalnet towards it. Maybe it makes more sense to keep the two completely separate?

**Comment by niksirbi** [comment]
@b-peri takeaways from today's chat on this:

- do the filtering only on "pose_tracks", not on "confidence" and (in the future) also not on "velocity" or other derivative variables. Basically we want people to clean first, analyse later.
- drop the inplace idea for good, return a new dataset.
- the logging and reporting should probably have been different PRs (this is mostly my fault for shoving these features in here). Feel free to merge just the filtering bit and deal with logging and/or reporting separately. For this particular PR, since you have anyway done most of the work already, also feel free to merge including these bits, but if one of them becomes tedious, drop it, merge without it, and open a separate issue for the missing bit.

File: movement/analysis/filtering.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    }
+    ds_interpolated.attrs["log"].append(log_entry)
+
+    if inplace:
+        ds["pose_tracks"] = ds_interpolated["pose_tracks"]
+        ds["confidence"] = ds_interpolated["confidence"]
+        ds.attrs["log"] = ds_interpolated.attrs["log"]
+        return None
+    else:
+        return ds_interpolated
```
[the remark this answers is not in the document]
Comment (b-peri) [reply movement/analysis/filtering.py]: I agree completely! I've now removed the `inplace` parameters from both functions.

File: movement/analysis/filtering.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+    log_entry = {
+        "operation": "interp_pose",
+        "method": method,
+        "limit": limit,
+        "max_gap": max_gap,
+        "inplace": inplace,
+        "datetime": str(datetime.now()),
+    }
+    ds_interpolated.attrs["log"].append(log_entry)
```
[the remark this answers is not in the document]
Comment (b-peri) [reply movement/analysis/filtering.py]: Done! Many thanks for the example code!

File: movement/analysis/filtering.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (b-peri) [reply movement/analysis/filtering.py]: I agree with this. I've now changed the function name and removed the `interpolate` parameter from this function altogether.

File: movement/analysis/filtering.py (hunk 1 above)
Comment (b-peri) [reply movement/analysis/filtering.py]: Thanks for the heads up here! The diagnostics block should now be fixed!

File: movement/filtering.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            )
+            n_points = ds.time.values.shape[0]
+            prop_nans = round((n_nans / n_points) * 100, 1)
+            diagnostic_report += (
+                f"\n   {kp}: {n_nans}/{n_points} ({prop_nans}%)"
+            )
+
+    # Write diagnostic report to logger
+    logger = logging.getLogger(__name__)
+    logger.info(diagnostic_report)
```
Comment (b-peri) [inline movement/filtering.py:63]: I'm currently struggling a bit to see how this approach is more convenient for the user than simply printing the diagnostics, as we did previously. However, I admit that this may just be because I'm unfamiliar with Python logging! What is the expected interface here? How exactly is the user meant to access these logs when they want to? Would it really not be easier to simply allow user to toggle printing of the diagnostics on and off?

File: movement/filtering.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+from copy import copy
+from datetime import datetime
+from functools import wraps
+from typing import Union
+
+import numpy as np
+import xarray as xr
+
+
+def log_to_attrs(func):
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/filtering.py:11]: I think for now it's fine to leave as is. If we find ourselves re-using it in other modules, then we can think about refactoring it out.

File: movement/filtering.py (hunk 6 above)
Comment (niksirbi) [reply movement/filtering.py:63]: This is natural question to ask if you haven't worked with logging before (as I myself hadn't a year ago). Actually logging is made to handle both the writing to log files and printing to the terminal (this can be configured).
The way I understand it, logging has several advantages to just printing:
- there are different [logging levels](https://docs.python.org/3/library/logging.html#logging-levels), ranging in severity. These means that messages can be "annotated" so that others know whether sth is e.g. a warning or just some useful info.
- logging can handle both printing to terminal (like "print") as well as writing to text files. Moreover, what it prints is not just the message, but also relevant context information about it, e.g datetime and which code line trigerred the error.
  ```console
  2023-04-28 12:32:09,506 - ERROR - MainProcess validators.py:29 - File not found: .data/DLC_sample_poses.csv
  ```
  What information is included in the log can be configured by us. Needless to say this is very valuable for the users and for us during debugging.
- the different logging levels allow us to configure handling logic for them. For example, we can choose to log all levels to the log file (incl. DEBUG) and only show INFO, WARNING, and ERROR in the terminal. This means that we can still keep lots of granular information useful for debugging, without bombarding the user with countless messages.

I hope this is clearer now. FYI you can find the movement logs in `~/.movement/movement.log`. Logs generated during testing are sent to a different, temporary location.

**Review by niksirbi — APPROVED** [review]
Hey @b-peri, thanks a lot for implementing the suggestions!

I made some final tweaks:
- I merged the main branch to resolve conflicts
- I renamed the `filter_diagnostics` function to `report_nan_values`, to more explicitly reflect its purpose. I also changed the way it's invoked by filtering/interpolation functions, so that the user gets an overview of the change in NaNs (made by the operation).
- I switched to using `xarray's built-in `isnull()` and `copy` methods (instead of `numpy.isnan` and `copy.copy`
- I rephrased some of the docstrings
- I added `fill_value="extrapolate"` to the `interpolate_na` method, to also take care of nan values at the start and end of the array (if they are smaller than the `max_gap`).
- I added the new functions to the API index in the docs
- I wrote a new example for the docs, showcasing how the new functions can be used to clead data

I will merge this as is now.

I think we should add some more extensive testing to cover some edgecases, but I will open a separate issue for that.
